lnwallet/fee_estimator.go

```diff
@@ -518,3 +543,206 @@ func (s SparseConfFeeSource) ParseResponse(r io.Reader) (map[uint32]uint32, erro
+// confirmation and returns the estimated fee expressed in sat/kw.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) EstimateFeePerKW(numBlocks uint32) (SatPerKWeight, error) {
+	if numBlocks > maxBlockTarget {
+		numBlocks = maxBlockTarget
+	} else if numBlocks < minBlockTarget {
+		return 0, fmt.Errorf("conf target of %v is too low, minimum "+
+			"accepted is %v", numBlocks, minBlockTarget)
+	}
+
+	feePerKb, err := w.getCachedFee(numBlocks)
+	if err != nil {
+		return 0, err
+	}
+
+	// If the result is too low, then we'll clamp it to our current fee
+	// floor.
+	satPerKw := SatPerKVByte(feePerKb).FeePerKWeight()
+	if satPerKw < FeePerKwFloor {
+		satPerKw = FeePerKwFloor
+	}
+
+	walletLog.Debugf("Web API returning %v sat/kw for conf target of %v",
+		int64(satPerKw), numBlocks)
+
+	return satPerKw, nil
+}
+
+// Start signals the FeeEstimator to start any processes or goroutines it needs
+// to perform its duty.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Start() error {
+	var err error
+	w.started.Do(func() {
+		walletLog.Infof("Starting web API fee estimator")
+
+		w.updateFeeTimer = time.NewTimer(w.randomFeeUpdateTimeout())
+		w.updateFeeEstimates()
+
+		w.wg.Add(1)
+		go w.feeUpdateManager()
+
+	})
+	return err
+}
+
+// Stop stops any spawned goroutines and cleans up the resources used by the
+// fee estimator.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Stop() error {
+	w.stopped.Do(func() {
+		walletLog.Infof("Stopping web API fee estimator")
+
+		w.updateFeeTimer.Stop()
+
+		close(w.quit)
+		w.wg.Wait()
+	})
+	return nil
+}
+
+// RelayFeePerKW returns the minimum fee rate required for transactions to be
+// relayed.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) RelayFeePerKW() SatPerKWeight {
+	return FeePerKwFloor
+}
+
+// randomFeeUpdateTimeout returns a random timeout between minFeeUpdateTimeout
+// and maxFeeUpdateTimeout that will be used to determine when the Estimator
+// should retrieve fresh fees from its API.
+func (w *WebAPIFeeEstimator) randomFeeUpdateTimeout() time.Duration {
+	lower := int64(minFeeUpdateTimeout)
+	upper := int64(maxFeeUpdateTimeout)
+	return time.Duration(prand.Int63n(upper-lower) + lower)
+}
+
+// getCachedFee takes in a target for the number of blocks until an initial
+// confirmation and returns an estimated fee (if one was returned by the API). If
+// the fee was not previously cached, we cache it here. Note that the feesMtx
+// should be held while calling this function, to avoid a race condition.
+func (w *WebAPIFeeEstimator) getCachedFee(numBlocks uint32) (uint32, error) {
+	w.feesMtx.Lock()
+	defer w.feesMtx.Unlock()
+	for target := numBlocks; target >= minBlockTarget; target-- {
+		fee, ok := w.feeByBlockTarget[target]
+		if !ok {
+			continue
+		}
+
+		w.feeByBlockTarget[numBlocks] = fee
```

Comment:
We only need to override the entry if it wasn't already found in the map.

Reply:
fixed